### ISSUE TITLE
Add support for Promises

### ISSUE THREAD

**Issue by aydrian** [SparkPost/node-sparkpost#107]
Looks like using the bluebird library will give us the ability to support both promises and callbacks.

**Comment by thorinisme** [comment]
you don't need bluebird. Promises are now native. You just need to return a promise if there is no callback. Exemple : 

`module.exports = function(client) {

  return {
    send: function (options, callback) {
      if (!callback) {
        return new Promise(function(resolve, reject) {
          //do assync the job
          // execute resolve(result); if it's ok
          // execute reject(err); if it fail
        });
      }
    }
  }
}`

**Comment by aydrian** [comment]
Hi. Thanks for the feedback. It was my understanding that Promises are native as of `4.0`. We still support versions below that. It was also my understanding that bluebird still performs better than the native implementation. Open to thoughts and ideas.
We are still supporting the following NodeJS versions from our (.travis.yml)
```
node_js:
- '4.1'
- '4.0'
- '0.12'
- '0.10'
```

**Comment by thorinisme** [comment]
Hi, i suggest you to read this benchmark:
http://ciplogic.com/index.php/blog/92-native-vs-bluebird-vs-core-promise-benchmark

Node full support Promise since V4. You can add Promise check: `if (Promise) {} else {} ` to be compatible with oldest version.

Even if user use recent version of node, some of them could be intersted in use the old fashion way (with a callback).

I suggest you to support both. If a callback is pass in parameter, then use the old fashion way, OOtherwise if promise is implemented, return a promise. Otherwise throw an exception. That way you keep you current function signature.

exemple for "send" method in "transmissions.js':

```node
function send(options, callback) {
  // put here your previous code
  throw new Error('pouet');
}

return = {
  send: function(options, callback) {
    if (callback) {
      return send(options, callback);
    }
    if (Promise) {
      return new Promise(function(resolve, reject) {
        return send(options, resolve);
      });
    }
    else {
      throw new Error('Promise are not supported yet by your node version. Please provide a callback to the function.');
    }
  }
}
```

**Comment by aydrian** [comment]
Yes. I was intending to support both callbacks and Promises.
So you're suggestion I do something like 
```javascript
var Promise = Promise || require('bluebird');
```
To use native promises for 4.0+ and bluebird if < 4.0?

**Comment by thorinisme** [comment]
yes you can do this

**Comment by aydrian** [comment]
Thank you. This will be my next PR I work on for 2.0.0. I welcome your feedback when I get it submitted.

**Comment by thorinisme** [comment]
i edited code in previous message (i have been using git for a few time)

**Comment by jasonrhodes** [comment]
Hmm, I'm not sure that `send(options, resolve)` would work because `resolve` would get called with an error if something went wrong. 

The other issue we have is that some of our functions do early error checking and call the callback in multiple places, which means we would have to check for the presence of a callback in multiple places and then the code gets pretty unwieldy. 

`q` has `nbind` and `nfbind` methods (previously called `denodeify`) that take care of a lot of this work, and `bluebird` has similar methods as well. But mixing a dep as big as bluebird or q with native promises might make it hard to debug issues.

Another way around this is to use https://github.com/stefanpenner/es6-promise to polyfill native promises, and then use something like https://github.com/matthew-andrews/denodeify to do our denodeifying one way for everyone. Then we ship a separate file that wraps all of our methods like that and ask promise users to do `require('sparkpost').promiseClient`

**Comment by thorinisme** [comment]
```
Promise = Promise || require('bluebird');
function send(options, callback) {
  // put here your previous code
}
module.exports = function(client) {
  return = {
    send: function(options, callback) {
      if (callback) {
        return send(options, callback);
      }
      return new Promise(function(resolve, reject) {
        send(options, function(err, res) {
          if (err) {
            reject(err);
          } else {
            resolve(res);
          }
        });
      });
    }
  }
}
```

**Comment by bradennapier** [comment]
Why not just include this by default so one can do

```
return Sparkpost.transmissions.send(email)
     .then( ..... )
```

without all the extra steps?

**Comment by aydrian** [comment]
Closed by #154
